src/bindings/fluid_ladspa.h

```diff
@@ -21,220 +21,129 @@
+    LADSPA_Descriptor_Function descriptor;
 
-/* Purpose:
- * Frees all memory and shuts down the Fx block.
- * The synthesis thread must be stopped, when calling.
- */
-void fluid_LADSPA_shutdown(fluid_LADSPA_FxUnit_t* FxUnit);
+} fluid_ladspa_lib_t;
 
+typedef struct _fluid_ladspa_port_state_t
+{
+    int num_inputs;
+    int num_outputs;
 
+} fluid_ladspa_port_state_t;
 
-/*
- * fluid_handle_LADSPA_XXX
- * Those functions are called from fluid_cmd, when a command is entered on the command line.
- */
+typedef struct _fluid_ladspa_plugin_t
+{
+    /* plugin instance id unique to the synth */
+    int id;
 
-/* Purpose:
- * - Resets LADSPA Fx unit to bypass.
- * - Removes all plugins from the reverb unit.
- * - Releases all libraries.
- * Note: It would be more efficient to keep the libraries. But then the user would have to restart fluidsynth each time
- * a plugin is recompiled.
- */
-int fluid_LADSPA_handle_clear(fluid_synth_t* synth, int ac, char** av, fluid_ostream_t out);
-
-/* Purpose:
- * Loads the plugins added with 'ladspa_add' and then start the Fx unit.
- * Internal processes:
- * - load the LADSPA plugin libraries
- * - instantiate the plugins
- * - connect the plugins
- * - set the bypass switch to 'not bypassed'
- */
-int fluid_LADSPA_handle_start(fluid_synth_t* synth, int ac, char** av, fluid_ostream_t out);
+    const LADSPA_Descriptor *desc;
+    LADSPA_Handle *handle;
 
-/* Purpose:
- * Adds one plugin into the list of the LADSPA Fx unit.
- * This is only allowed, while the Fx block is in 'bypass' state (after clear).
- */
-int fluid_LADSPA_handle_add(fluid_synth_t* synth, int ac, char** av, fluid_ostream_t out);
+    volatile int active;
 
-/* Purpose:
- * Declares a user control node and a value; for further processing in ladspa_start.
- */
-int fluid_LADSPA_handle_declnode(fluid_synth_t* synth, int ac, char** av, fluid_ostream_t out);
+    /* Used to keep track of the port connection states */
+    fluid_ladspa_port_state_t *ports;
 
-/* Purpose:
- * Assigns a value to the a user control node
- */
-int fluid_LADSPA_handle_setnode(fluid_synth_t* synth, int ac, char** av, fluid_ostream_t out);
+} fluid_ladspa_plugin_t;
 
-#endif /* LADSPA */
+typedef struct _fluid_ladspa_node_t
+{
+    char *name;
+    fluid_ladspa_node_type_t type;
+    LADSPA_Data *buf;
+
+    char num_inputs;
+    char num_outputs;
+
+} fluid_ladspa_node_t;
+
+typedef struct _fluid_ladspa_fx_t
+{
+    int sample_rate;
+
+    int audio_groups;
+    int effects_channels;
+    int audio_channels;
+
+    fluid_ladspa_lib_t *libs[FLUID_LADSPA_MAX_LIBS];
+    int num_libs;
+
+    fluid_ladspa_node_t *nodes[FLUID_LADSPA_MAX_NODES];
+    int num_nodes;
+
+    /* plugins are really plugin instances */
+    fluid_ladspa_plugin_t *plugins[FLUID_LADSPA_MAX_PLUGINS];
+    int num_plugins;
+
+    /* used to generate the unique plugin ids */
+    int next_plugin_id;
+
+    fluid_rec_mutex_t api_mutex;
 
-#endif  /* _FLUID_LADSPA_H */
+    volatile int state;
```

Comment:
why volatile?

Reply:
To ensure that the state variable is guaranteed to be written before the API mutex is released. Otherwise there could be a race condition, leading to two command handlers activating or deactivating the effects engine more than once.